apps: Migrate from ODFE to OpenSearch

aarnq opened the pull request:
**What this PR does / why we need it**:
This migrates from ODFE to OpenSearch, as the OpenSearch project replaces ODFE.

**Which issue this PR fixes**:
fixes #588
fixes #726 

**Public facing documentation PR**: -
<!-- https://github.com/elastisys/compliantkubernetes/pull/ -->

**Special notes for reviewer**:

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [x] requires running a migration script.
    - [x] will create noticeable cluster degradation.
          - At least some during migration Kibana and Elasticsearch -> OpenSearch and OpenSearch Dashboards.
    - [ ] requires draining and/or replacing nodes.
    - [x] will change any APIs.
          - `elasticsearch` -> `opensearch`
          - `kibana` -> `opensearch.dashboards`
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [ ] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [x] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config**:
Updated.
<!--
Here are the commit prefixes and comments on when to use them:
all: (things that touch on more than one of the areas below, or don't fit any of them)
apps: (changes to the applications running in both/all clusters)
apps sc: (changes to applications in the service cluster)
apps wc: (changes to applications in the workload cluster)
docs: (documentation)
tests: (test related changes)
pipeline: (the pipeline)
config: (configuration, e.g. add/remove/rename a parameter, this is not for changes to the default values for an application that would go into `apps [sc/wc]`)
bin: (changes to binaries or scripts used manage ck8s)
release: (anything release related)

Example commit prefix usage:

git commit -m "docs: Add instructions for how to do x"
-->

aarnq commented:
I missed an important detail, OpenSearch with the default index storage uses mapped memory. And they check against the  system setting `vm.max_map_count=262144` which is most likely not set as the default of the Kubernetes nodes.
It is possible to [set this](https://opensearch.org/docs/latest/opensearch/install/important-settings/) by ssh:ing into the nodes, but I'll investigate if there is another solution.

Xartos reviewed (commented):
I have not gone through everything yet. But I'll dump the comments I have for now, will probably continue tomorrow

aarnq commented on config/config/sc-config.yaml:
```diff
@@ -257,40 +258,86 @@ dex:
-  # Single-sign-on using OIDC
-  # NOTE: SSO using OIDC requires LetsEncrypt Production
+
+  dashboards:
```
The Kibana fork is called OpenSearch Dashboards, so I figured it would be easiest this way to keep it consistent.

aarnq commented on config/config/wc-config.yaml:
```diff
@@ -92,16 +92,9 @@ falco:
-    count: 1
+## OpenSearch subdomain for forwarding logs.
+opensearch:
+  subdomain: opensearch
```
True, I thought so to but I guess I forgot it along the way.

aarnq commented on helmfile/50-applications.yaml:
```diff
@@ -267,15 +254,155 @@ releases:
 
+# OpenSearch
+- name: opensearch-secrets
+  namespace: elastic-system
```
We could do that, will look into it.

aarnq replied:
It is mostly done I think, will test it when I fix the migration steps but I need to set up a new OpenDistro source.

Xartos commented on migration/v0.18.x-v0.19.x/opensearch-migration-run.sh:
```diff
@@ -0,0 +1,289 @@
+
+echo "--- Delete ODFE. (Skip if you want to keep it for now, remove it now if you are going to reuse the subdomains.)"
+echo "ck8s ops helmfile sc -l group=opendistro delete"
+echo -n "- run? [y/s/N]: "
```
It should be fine, but I would like some way of clarifying that `s` is the skip option.
So maybe change to something like this:

```suggestion
echo -n "- run? [y/s (skip)/N]: "
```

It's not super pretty, so if you have any other idea

aarnq replied:
I've added it into the first line, so now it says "Skip using option 's' if..."

Xartos commented on migration/v0.18.x-v0.19.x/opensearch-migration-run.sh:
```diff
@@ -0,0 +1,289 @@
+echo "---"
+echo
+
+echo "--- Deploy OpenSearch."
```
I really like the prompts in the script. But I feel like the non destructive once should be fine to just execute without it.

aarnq replied:
Done. 
Saw I had double steps for some things which should not be there so I fixed that as well.

Xartos commented on migration/v0.18.x-v0.19.x/upgrade-apps.md:
```diff
@@ -25,3 +49,9 @@
+
+    Run script `opensearch-migration-clean.sh`, this will delete deprecated parameters for ODFE in `sc-config.yaml` and `wc-config.yaml`.
+
+    Delete deprecated secrets for `elasticsearch` in `secrets.yaml`.
```
Wouldn't it be nice to have this in the script as well?

aarnq replied:
I'll see if I can add it.

Xartos replied:
For inspiration, this is how we have done it previously
https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration/v0.7.x-v0.8.x/migrate-object-storage.sh#L42-L46

Xartos replied:
Did you look into this anything yet?

aarnq replied:
No, I haven't gone through the whole `migration-run` script yet, but when I do I'll add it.

Xartos commented on migration/v0.18.x-v0.19.x/opensearch-migration-clean.sh:
```diff
@@ -0,0 +1,42 @@
+
+set -euo pipefail
+
+IFS=$'\n'
```
The reason for this? I don't see any arrays being created, but maybe there's something else that uses this.

aarnq replied:
I think I just used one as a template for the others.

Xartos replied:
Right, as long as it works I'm fine with keeping it.

Xartos commented on helmfile/values/prometheus-alerts-wc.yaml.gotmpl:
```diff
@@ -1,6 +1,5 @@
 # Note: These values are used for the wc reader. I.e. **in SC** despite the name
 
-esNodeCount: {{ add .Values.elasticsearch.dataNode.count .Values.elasticsearch.clientNode.count .Values.elasticsearch.masterNode.count }}
```
Why don't we need this one anymore? To be honest, I'm not really sure why we need them right now but I know there's a reason

aarnq replied:
Because the rule for it is disabled just below it, but I can double check it in case I've missed something.
```yaml
defaultRules:
  rules:
    elasticsearch: false
```

aarnq replied:
Have confirmed, this rule does not get created.

Xartos commented on helmfile/values/opensearch/slm.yaml.gotmpl:
```diff
@@ -0,0 +1,26 @@
+{{ if not (or (eq .Values.objectStorage.type "s3") (eq .Values.objectStorage.type "gcs") ) }}
```
I don't see us using these variables in here, do we still need the check for this file?

aarnq replied:
It is carried over from elasticsearch-slm, I guess the check makes sense since it requires that OpenSearch is set up with a snapshot repository, which we only do with object storage.

Xartos commented on helmfile/values/opensearch/secrets.yaml.gotmpl:
```diff
@@ -0,0 +1,51 @@
+opensearch:
```
Shouldn't we have the objectstorage type, check here instead? Or does it work without one?

aarnq replied:
It works without one if it runs without snapshots.

Xartos commented on helmfile/values/opensearch/curator.yaml.gotmpl:
```diff
@@ -0,0 +1,24 @@
+  accountSecret: opensearch-curator-user
+  clusterEndpoint: {{ .Values.opensearch.clusterName }}-master:9200
+
+startingDeadlineSeconds: 600
```
Maybe not part of the issue, but I would like to move this to the config file. Objections for that?

Xartos replied:
Yes, basically the same as this https://github.com/elastisys/compliantkubernetes-apps/blob/54f1d1ab2434ebc4b19c163c373002e324623d66/helmfile/values/opensearch/slm.yaml.gotmpl#L9

aarnq replied:
Added it, I also looked through and made the same across all cronjobs.

Xartos reviewed (commented):
Some more comments

robinAwallace reviewed (commented):
Great work! Did not look through everything but here are some comments :slightly_smiling_face:

aarnq commented on WIP-CHANGELOG.md:
```diff
@@ -13,9 +13,26 @@
 ### Fixed
 
+- Fluentd will now properly detect changes in container logs.
+- The `init` script will now properly generate secrets for new configuration options.
```
No, if new properties were added they where given the standard `somelongsecret` as set in the `config/secrets/*` files.
So this is not just for opensearch, but it is the first new addition since secret generation was integrated into `init`. :slightly_smiling_face:

robinAwallace replied:
Ah okay then skip my comment here then :slightly_smiling_face:

aarnq commented:
I see now that I need to change the migration steps to include applying the policies to the restored indicies, as this is not done automatically.

robinAwallace commented on helmfile/values/fluentd-wc.yaml.gotmpl:
```diff
@@ -93,7 +93,6 @@ extraConfigMaps:
       tag kubeaudit.*
       read_from_head true
       skip_refresh_on_startup true
-      enable_watch_timer false
```
Is this not needed anymore for fluentd?

aarnq replied:
I had troubles getting logs for the Kubernetes index with it disabled. It would get some logs when it started, but afterwards it would basically stop sending logs.
Without the watch timer it uses ionotify to check for update which apparently does not work reliably for me.

robinAwallace commented on migration/v0.18.x-v0.19.x/opensearch-migration-run.sh:
```diff
@@ -0,0 +1,252 @@
+echo "---"
+echo
+
+echo "--- Delete ODFE. (Skip using option 's' if you want to keep it for now or already done so.)"
+echo "ck8s ops helmfile sc -l group=opendistro delete"
+echo -n "- run? [y/s/N]: "
+read -r reply
+if [[ "${reply}" == "y" ]]; then
+    "${here}/../../bin/ck8s" ops helmfile sc -l group=opendistro delete
+elif [[ "${reply}" == "s" ]]; then
+    echo "skipping"
+else
+    exit 1
+fi
+echo "---"
+echo
+
+echo "--- Delete ODFE. (Skip using option 's' if you want to keep it for now or already done so.)"
+echo "ck8s ops helmfile sc -l group=opendistro delete"
+echo -n "- run? [y/s/N]: "
+read -r reply
+if [[ "${reply}" == "y" ]]; then
+    "${here}/../../bin/ck8s" ops helmfile sc -l group=opendistro delete
+elif [[ "${reply}" == "s" ]]; then
+    echo "skipping"
+else
+    exit 1
+fi
+echo "---"
+echo
```
Doing the same operation two times.

aarnq replied:
Fixed.

viktor-f commented on helmfile/values/opensearch/slm.yaml.gotmpl:
```diff
@@ -0,0 +1,26 @@
+{{ if not (or (eq .Values.objectStorage.type "s3") (eq .Values.objectStorage.type "gcs") ) }}
+{{ fail "\nERROR: Elasticsearch slm requires s3 or gcs object storage, see Values.objectStorage.type" }}
```
```suggestion
{{ fail "\nERROR: Opendistro slm requires s3 or gcs object storage, see Values.objectStorage.type" }}
```

viktor-f reviewed (commented):
Really well done! I think I have looked through all your changes.

Will you adopt the new retention config that you have here: https://github.com/elastisys/compliantkubernetes-apps/commit/f1a96f2bf35e500f67401cab1a0abaf4ed2d28a0

aarnq commented on helmfile/values/opensearch/backup.yaml.gotmpl:
```diff
@@ -1,7 +1,18 @@
+
+# Ignore security plugin index as per
+# https://opendistro.github.io/for-elasticsearch-docs/docs/elasticsearch/snapshot-restore/#security-plugin-considerations
+indices: "*,-.opendistro_security"
```
Yes, it is still done so in the plugin. I'll update the link to OpenSearch documentation.

aarnq commented on helmfile/charts/opensearch/curator/files/action_file.yml:
```diff
@@ -15,14 +15,14 @@ actions:
+      disk_space: {{ .Values.retention.other_gb }}
       use_age: True
       source: creation_date
     - filtertype: kibana
```
Yes it'll exclude `.kibana*` indices, however I'm looking into using `.opensearch_dashboards*` and if that works I'll update it.

aarnq replied:
This is done, it also uses the new format for retention. Though it isn't exposed in the config.

viktor-f reviewed (commented):
Looks good to me. Is there some changes left to do or something that should be tested?

(as a general note for discussion: I think I would prefer separate fix commits on PRs that are this large, then squashing everything before merging to main. Otherwise it is a bit hard to see what was changed in large files, e.g. `sc-config.yaml`)

aarnq commented:
@viktor-f Yeah, there are some small changes left addressing the comments above. 
And I agree, I've kind of learnt that with this one.

It might also be best that we waited for the helm chart to be updated with OS 1.2.2, or as it looks 1.2.3 to be released with mitigations for the third CVE on log4j.

OlleLarsson commented on config/config/sc-config.yaml:
```diff
@@ -259,40 +260,86 @@ dex:
-    # Where to find roles
-    roles_key: groups
-    # Scope - add 'groups' if groups claim is supported
+    subject_key: email # Where to find subject.
+    roles_key: groups  # Where to find roles.
```
Let's change to
```suggestion
    subjectKey: email # Where to find subject.
    rolesKey: groups  # Where to find roles.
```

OlleLarsson commented on config/config/sc-config.yaml:
```diff
@@ -259,40 +260,86 @@ dex:
+      definition:
+        users:
+          - set-me # Developer Name
+    - mapping_name: kibana_user
```
Maybe? Or have they kept the `kibana_user` in the defaults? 
```suggestion
    - mapping_name: dashboards_user
```

aarnq replied:
Yeah, `kibana_user` is still a static role, the same with `kibana_server`.

OlleLarsson commented on helmfile/values/prometheus-alerts-sc.yaml.gotmpl:
```diff
@@ -1,8 +1,8 @@
-eskubernetesSizeGB: {{ .Values.elasticsearch.curator.retention.kubernetesSizeGB }}
-esotherSizeGB: {{ .Values.elasticsearch.curator.retention.otherSizeGB }}
-esauthLogSizeGB: {{ .Values.elasticsearch.curator.retention.authLogSizeGB }}
+esNodeCount: {{ add .Values.opensearch.dataNode.count .Values.opensearch.clientNode.count .Values.opensearch.masterNode.count }}
+eskubeAuditSizeGB: {{ .Values.opensearch.curator.retention.kubeAuditSizeGB }}
+eskubernetesSizeGB: {{ .Values.opensearch.curator.retention.kubernetesSizeGB }}
+esotherSizeGB: {{ .Values.opensearch.curator.retention.otherSizeGB }}
+esauthLogSizeGB: {{ .Values.opensearch.curator.retention.authLogSizeGB }}
```
Use `os` as shot notation for opensearch?
```suggestion
osNodeCount: {{ add .Values.opensearch.dataNode.count .Values.opensearch.clientNode.count .Values.opensearch.masterNode.count }}
osKubeAuditSizeGB: {{ .Values.opensearch.curator.retention.kubeAuditSizeGB }}
osKubernetesSizeGB: {{ .Values.opensearch.curator.retention.kubernetesSizeGB }}
osOtherSizeGB: {{ .Values.opensearch.curator.retention.otherSizeGB }}
osAuthLogSizeGB: {{ .Values.opensearch.curator.retention.authLogSizeGB }}
```

OlleLarsson commented on helmfile/values/opensearch/secrets.yaml.gotmpl:
```diff
@@ -0,0 +1,51 @@
+  {{- end }}
+  {{- end }}
+
+  accounts:
```
or internalUsers?
```suggestion
  users:
```

viktor-f commented:
I'm guessing this would be easy to fix in this PR: https://github.com/elastisys/compliantkubernetes-apps/issues/726
(setting resources for slm job)

aarnq commented:
Fixes #726

OlleLarsson commented on config/config/wc-config.yaml:
```diff
@@ -92,17 +92,6 @@ falco:
 
-## Elasticsearch cluster topolgy.
-## Used in prometheus alerts.
-elasticsearch:
```
No longer used or?

aarnq replied:
It was used to get the node count into wc/user alerts, but the alerts was turned off.

Xartos reviewed (approved):
LGTM! Just remember to squash some commits before merging :)
